File: Sorts/SelectionSort/SelectionSort.c

//=====INCLUDE=====//
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include "../Utils/Generator.h"

//=====DEFINE=====//
#define LEN 100

//=====PROTOTYPES=====//
void selectionSorti(int arr[], int first, int last);
void swapi(int arr[], int a, int b);

void selectionSortd(double arr[], int first, int last);
void swapd(double arr[], int a, int b);

//=====MAIN=====//
int main(char argc, char *args) {
    double arr[LEN];
    for(int i = 0; i < 100; i++) {
        generated(arr, LEN, 10000);
        selectionSortd(arr, 0, LEN);
        printf("%i\n", verifyd(arr, LEN));
    }
    return 0;
}

//=====METODOS=====//
void selectionSorti(int arr[], int first, int last) {
    int min;
    for(int i = first; i < last - 1; i++) {
        min = i;
        for(int j = i + 1; j < last; j++) 
            if(arr[j] < arr[min]) min = j;
        swapi(arr, i, min);
    }
}

void swapi(int arr[], int a, int b) {
    int aux = arr[a];
    arr[a] = arr[b];
    arr[b] = aux;
}

void selectionSortd(double arr[], int first, int last) {
    int min;
    for(int i = first; i < last - 1; i++) {
        min = i;
        for(int j = i + 1; j < last; j++) 
            if(arr[j] < arr[min]) min = j;
        swapd(arr, i, min);
    }
}

void swapd(double arr[], int a, int b) {
    double aux = arr[a];
    arr[a] = arr[b];
    arr[b] = aux;
}

File: Sorts/SelectionSort/SelectionSort.java
package Sorts.SelectionSort;

import Sorts.Utils.Generator;

public class SelectionSort {
    
    public static void main(String[] args) {
        double[] arr = new double[100];
        
        for(int i = 0; i < 100; i++) {
            Generator.generate(arr, 10000);
            selectionSort(arr, 0, arr.length);
            System.out.println(Generator.verify(arr)); 
        }
    }

    public static void selectionSort(int[] arr, int first, int last) {
        int min;
        for(int i = first; i < last - 1; i++) {
            min = i;
            for(int j = i + 1; j < last; j++) 
                if(arr[j] < arr[min]) min = j;
                
            if(i != min) swap(arr, i, min);
        }
    }

    public static void swap(int[] arr, int a, int b) {
        int aux = arr[a];
        arr[a] = arr[b];
        arr[b] = aux;
    }

    public static void selectionSort(double[] arr, int first, int last) {
        int min;
        for(int i = first; i < last - 1; i++) {
            min = i;
            for(int j = i + 1; j < last; j++)
                if(arr[j] < arr[min]) min = j;
            
            if(i != min) swap(arr, i, min);
        }
    }

    public static void swap(double[] arr, int a, int b) {
        double aux = arr[a];
        arr[a] = arr[b];
        arr[b] = aux;
    }
}
